File: 3-2_fractKnapsack/fractKnapsackCpp/fractional_knapsack.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <iomanip>

// using std::vector;
using namespace std;


// Input: capacity - total weight knapsack holds
// Input: values - value of each item
// Input: weights - weight of each item
// Output: value - max value of items that fit in knapsack
double get_optimal_value(int capacity, vector<int> weights, vector<int> values) {
  double value = 0.0;
  int num_items = weights.size();
  vector< pair <double, int> > mypair(num_items);
  
  // Calc weight value and store in pair vector
  for (int i = 0; i < num_items; ++i) {
    mypair[i].first = (double)values[i] / (double)weights[i];
    mypair[i].second = weights[i];
  }  
  
  // Sort pairs in desc order by value
  std::sort(mypair.begin(), mypair.end());
  
  // Accumulate values for available capacity
  for (int i = num_items - 1; i > -1; --i) {
    if (mypair[i].second <= capacity) {
      value = value + (mypair[i].first * mypair[i].second);
      capacity = capacity - mypair[i].second;
      if (capacity < 0)
        capacity = 0;
    } else {
      value = value + (mypair[i].first * capacity);
      capacity = 0;
    }
  }
  
/* 
  // Debug  state: view inputs desc sort by value 
  for (int i = num_items - 1; i > -1; --i) {
    std::cout << fixed << setprecision(4) 
        << mypair[i].first << " " 
        << mypair[i].second << "\n";
  } 
*/ 


  return value;
}


int main() {
  int n;
  int capacity;
  std::cin >> n >> capacity;
  vector<int> values(n);
  vector<int> weights(n);
  for (int i = 0; i < n; i++) {
    std::cin >> values[i] >> weights[i];
  }

  double optimal_value = get_optimal_value(capacity, weights, values);

  //std::cout.precision(4);
  std::cout << fixed << setprecision(4) 
      << optimal_value << std::endl;
  return 0;
}

File: 3-2_fractKnapsack/fractKnapsackJava/FractionalKnapsack.java
import java.util.Scanner;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import javafx.util.Pair;

public class FractionalKnapsack {
  
    private static double getOptimalValue(int capacity, int[] values, int[] weights) {
        double value = 0;
        int num_items = values.length;
        ArrayList <Pair <Double, Integer> > mypair = 
                  new ArrayList <Pair <Double, Integer> > ();
                  
        // Calc weight value and store in pair vector          
        for (int i = 0; i < num_items; ++i) {
          mypair.add(new Pair <Double, Integer>((double)values[i]/weights[i], weights[i]));
        }
        
        // Sort pairs in desc order by value
        mypair.sort(new Comparator<Pair<Double, Integer>>() {
          @Override
          public int compare(Pair<Double, Integer> o1, Pair<Double, Integer> o2) {
            if (o1.getKey() > o2.getKey()) {
                return -1;
            } else if (o1.getKey().equals(o2.getKey())) {
                return 0; 
                          
            } else {
                return 1;
            }
          }
        });
        
        // Accumulate values for available capacity
        for (int i = 0; i < num_items; ++i) {
          if (mypair.get(i).getValue() <= capacity) {
            value = value + (mypair.get(i).getKey() * mypair.get(i).getValue());
            capacity = capacity - mypair.get(i).getValue();
            if (capacity < 0)
              capacity = 0;
          } else {
            value = value + (mypair.get(i).getKey() * capacity);
            capacity = 0;
          }
        }
        
        /*
        // debug: print values and weights
        for (int i = 0; i < num_items; ++i) {
          System.out.println(mypair.get(i).getKey());
        }
        */
        
        return value;
    }

    public static void main(String args[]) {
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();
        int capacity = scanner.nextInt();
        int[] values = new int[n];
        int[] weights = new int[n];
        for (int i = 0; i < n; i++) {
            values[i] = scanner.nextInt();
            weights[i] = scanner.nextInt();
        }
        System.out.println(getOptimalValue(capacity, values, weights));
    }
} 
